crates/core/src/agent/chain_store.rs

```diff
@@ -217,6 +184,16 @@ impl ChainStore {
     }
 }
 
+impl ContentStore for ChainStore {
+    fn content_storage(&self) -> Arc<RwLock<dyn ContentAddressableStorage>> {
```

Comment:
I don't see it as an improvement that we re-add a function that returns a reference to the lock. It enables creation of lock guards in arbitrary scopes which is what led to those dreaded deadlocks. Also, it enables calling contexts to write to the CAS with an immutable reference to the state.

Reply:
This function is private, so it's only available in this module, in the two trait implementations. So the guard cannot leak out. Does that change your outlook?

Reply:
Actually I want to look into if it is possible to access this by importing ChainStore...if this is publicly accessible then I think this is a problem.

Reply:
OK yeah, I misunderstood how trait methods work. It turns out that by importing `core::content_store::ContentStore`, you can then access this method, which would leak out the Mutex, which indeed is something we don't want to risk happening again.

One suggestion would be to remove ContentState, and have structs implement `GetContent` and `AddContent` explicitly. A little extra boilerplate, but offers proper encapsulation.

Reply:
@maackle Yeah I think that is the best option. It doesn't take much more code really but means we lose the awesome default implementations